crengine/src/private/lvfreetypeface.cpp

```diff
@@ -1801,27 +1825,35 @@ LVFontGlyphCacheItem *LVFreeTypeFace::getGlyph(lUInt32 ch, lChar32 def_char, lUI
                 rend_flags );             /* load flags, see below */
         if ( error == FT_Err_Execution_Too_Long && _hintingMode == HINTING_MODE_BYTECODE_INTERPRETOR ) {
             // Native hinting bytecode may fail with some bad fonts: try again with no hinting
+            CRLog::error("Font '%s': loading glyph too long!", _fileName.c_str());
             rend_flags |= FT_LOAD_NO_HINTING;
             error = FT_Load_Glyph( _face, ch_glyph_index, rend_flags );
         }
         if ( error ) {
             return NULL;  /* ignore errors */
         }
 
-        if (_embolden) {
-            FT_GlyphSlot_Embolden(_slot); // See setEmbolden() for details
+        if (_synth_weight > 0) {
+            if ( _slot->format == FT_GLYPH_FORMAT_OUTLINE ) {
+                // See setSynthWeight() for details
+                FT_Outline_Embolden(&_slot->outline, _synth_weight_strength);
+                FT_Outline_Translate(&_slot->outline, 0, -_synth_weight_half_strength);
+            }
         }
         if (_italic == 2) {
             FT_GlyphSlot_Oblique(_slot);
         }
-        if (_embolden || _italic==2) {
+        if (_synth_weight > 0 || _italic == 2) {
             // Render now that transformations are applied
             FT_Render_Glyph(_slot, _drawMonochrome?FT_RENDER_MODE_MONO:FT_RENDER_MODE_LIGHT);
         }
 
         item = newItem(&_glyph_cache, (lChar32)ch, _slot); //, _drawMonochrome
-        if (item)
+        if (item) {
+            if (item->origin_x < 0 && item->advance_26_6 == 0)
+                item->origin_x -= FONT_METRIC_TO_PX(_synth_weight_strength);
```

Comment:
I guess this section could also have your `Assume zero advance means it's a diacritic` comment?

~~Shouldn't this item->origin_x  fix be done in newItem(), where we actually compute it? This tweak here doesn't seem to depend on anything local to here.~~
OK, _synth_weight_strength is not available in newItem.
But it feel strange hacking fields of the thing returned that is supposed to be cached.
Can't we hack it before newItem, if the FT slot-bitmap is not readonly? This would look cleaner:

```diff
        if (_synth_weight > 0 || _italic == 2) {
            // Render now that transformations are applied
            FT_Render_Glyph(_slot, _drawMonochrome?FT_RENDER_MODE_MONO:FT_RENDER_MODE_LIGHT);
+           if (__synth_weight_strength != 0 && slot->bitmap_left < 0 && slot->metrics.horiAdvance == 0) {
+              slot->bitmap_left -= FONT_METRIC_TO_PX(_synth_weight_strength)
            }
        }
        item = newItem(&_glyph_cache, (lChar32)ch, _slot); 
```

Reply:
In general, I don't like the idea of modifying the glyph slot.

Reply:
OK, so may be just wrap this in an additional `if (_synth_weight > 0)`, so we know it's only relevant in this case?

```c
            item = newItem( &_glyph_cache, (lChar32)ch, _slot ); //, _drawMonochrome
            if (item) {
                if (_synth_weight > 0) {
                    // Assume zero advance means it's a diacritic: correct
                    // any origin_x for some reason you know better than me :)
                    if (item->origin_x < 0 && item->advance_26_6 == 0)
                        item->origin_x -= FONT_METRIC_TO_PX(_synth_weight_strength);
                }
                _glyph_cache.put( item );
            }
```

Reply:
Ok, give me some minutes...

Reply:
Thanks (I would have prefered my `if (_synth_weight > 0) {` instead of your` if (_synth_weight_strength != 0) {` - not for logical'ness, but for readability, as it's the same test we do above - so easier to skip when reading and not interested with synthetic stuff.)

Reply:
Am I right with this update of the comment (mainly about the fact we remove something _now_, because we'll be adding that *later* - or is this to counteract the FT_Outline_Translate?) ?
```c
                    // With synthetic weights, we'll add _synth_weight_strength to all
                    // advances. This should not be done with diacritics, so remove here
                    // by anticipation what we'll be adding later in the normal processing.
                    if (item->origin_x < 0 && item->advance_26_6 == 0) // probable diacritic
                        item->origin_x -= FONT_METRIC_TO_PX(_synth_weight_strength);
```

Reply:
> I would have prefered my `if (_synth_weight > 0) {` instead of your` if (_synth_weight_strength != 0) {` - not for logical'ness, but for readability, as it's the same test we do above - so easier to skip when reading and not interested with synthetic stuff.

If you set synth weight to 300 or to 600 and if you have only real 400 & 700 then `_synth_weight_strength` will be < 0.

Reply:
> Am I right with this update of the comment (mainly about the fact we remove something _now_, because we'll be adding that _later_ - or is this to counteract the FT_Outline_Translate?) ?

Now I'm trying to remember why I wrote this, but I can't. It seems the first variant... :)
Just remove or comment this fragment and test diacritic marks in simple FreeType renderer and synth weights.
`FT_Outline_Translate` we use only for Y axis.

Reply:
You added it last evening to fix diacritics with FreeType. It's probably the right thing to do (detecting here it's a diacritic, and fixing stuff).
But then ... I wonder if the synth_weight_stength you added last evening in measureText/DrawTextString couldn't have better be added here :)
(Although I'm quite ok with how measureText/DrawTextString ended up.)

Reply:
> But then ... I wonder if the synth_weight_stength you added last evening in measureText/DrawTextString couldn't have better be added here :)

It seemed to me that it was much more difficult to do it there, but there was already terribly confusing code there.

Reply:
> You added it last evening to fix diacritics with FreeType.

Oh no, it was a deep night :)

Reply:
:)

> If you set synth weight to 300 or to 600 and if you have only real 400 & 700 then _synth_weight_strength will be < 0.

But if I am right in thinking that what you do _here_ is to compensate what will be done _there_ in measureText/DrawTextString with:
`FT_Pos letter_spacing_26_6 = PX_TO_FONT_METRIC(letter_spacing) + _synth_weight_strength`

_There_ , you're not checking if < 0 , so you're reducing letter_spacing_26_6 when having a synthetic thinner font. My naive thinking is that there's no reason to not compensate for diacritics also in this case of thin'ning.

Reply:
> _There_ , you're not checking if < 0 , so you're reducing letter_spacing_26_6 when having a synthetic thinner font. My naive thinking is that there's no reason to not compensate for diacritics also in this case of thin'ning.

The width of the character above/below which the diacritical mark is located has changed, so the position of this mark must also change?
I seem to have remembered

Reply:
Regular (the e+^ is a bit off = font bug, but the _ is correctly centered below the o):

![image](https://user-images.githubusercontent.com/24273478/113485582-d2106c00-949d-11eb-8a9f-f345732532c4.png)


With your `_synth_weight_strength > 0`
```c
                if (_synth_weight > 0 && _synth_weight_strength > 0) {
                    // With synthetic weights, we'll add _synth_weight_strength to all
                    // advances. This should not be done with diacritics, so remove here
                    // by anticipation what we'll be adding later in the normal processing.
                    if (item->origin_x < 0 && item->advance_26_6 == 0) // probable diacritic
                        item->origin_x -= FONT_METRIC_TO_PX(_synth_weight_strength);
                }
```
![image](https://user-images.githubusercontent.com/24273478/113485592-dfc5f180-949d-11eb-99bf-9c7ef4c46add.png)
The ê seems OK (but then you corrected the font bug by luck :) - but the regularly correct _ is now no longer centered.

With `if (_synth_weight > 0) { // && _synth_weight_strength > 0)`:
![image](https://user-images.githubusercontent.com/24273478/113485632-17349e00-949e-11eb-973c-cea05c1f766a.png)
The ê font bug is preserved :) - and the _ seems more like with the regular.

Reply:
> With your _synth_weight_strength > 0

I do not have such


> ```c
>                     // With synthetic weights, we'll add _synth_weight_strength to all
>                     // advances. This should not be done with diacritics, so remove here
>                     // by anticipation what we'll be adding later in the normal processing.
> ```

Incorrect comment. In measureText/DrawText advance not added for diacritic mark.
Ok, I updated the last commit.

Reply:
> but then you corrected the font bug by luck

But I don't try to fix font bug at all, I was just fixing my bug of incorrect displacement of diacritics when using synthetic weight for FreeType rendering.

Reply:
OK, so I wrongly understood how this worked.
And OK, your last comment makes sense, I'm picking it.